lib/datadog/appsec/configuration/settings.rb

```diff
@@ -53,73 +52,71 @@ def self.add_settings!(base)
               end
 
               option :ruleset do |o|
-                o.default { ENV.fetch('DD_APPSEC_RULES', DEFAULT_APPSEC_RULESET) }
+                o.env 'DD_APPSEC_RULES'
+                o.default DEFAULT_APPSEC_RULESET
               end
 
               option :ip_denylist do |o|
-                o.default { [] }
+                o.type :array
+                o.default []
```

Comment:
There's a difference between block and value here:
1. `[]` would be `eval`'d when that line is evaluated, so ever produces a single array
2.`{ [] }` would be `eval`'d when the block is evaluated, so may or may not produce a single array

This led me to wonder if we should `freeze` that default in the 1. case to ensure it does not get mutated.

This then led me to wonder if we should in fact (transparently) `freeze` all configuration setting values, because once they are set they should not be mutated.

WDYT?

Reply:
I had to wrap it on `Core::Utils::SafeDup.frozen_or_dup(...)` because a test validates that when modifying the default value doesn't modify it by reference. 

```ruby
describe '#writer_options' do
  subject(:writer_options) { settings.tracing.test_mode.writer_options }

  it { is_expected.to eq({}) }

  context 'when modified' do
    it 'does not modify the default by reference' do
      settings.tracing.test_mode.writer_options[:foo] = :bar
      expect(settings.tracing.test_mode.writer_options).to_not be_empty
      expect(settings.tracing.test_mode.options[:writer_options].default_value).to be_empty
    end
  end
end
```

The options were like this before our changes:
```ruby
option :writer_options do |o|
  o.default { {} }
end
```

This is how they are after our changes:
```ruby
option :writer_options do |o|
  o.type :hash
  o.default({})
end
```

So I believe this was an oversight on my part in which using a default value as a block also meant that calling `default_value` would return a hash every time. To fix that particular spec, I had to use `Core::Utils::SafeDup.frozen_or_dup`

@lloeki suggested that we should be freezing the configuration value it gets set, so any modification to the configuration value should have to use `dup` first

Reply:
I think freezing instead of dup'ing seems reasonable as well, although I wonder if somewhere we may be doing something weird that doesn't like it (but perhaps in that situation it'd be worth fixing that code instead of not freezing)